### DISCUSSION TITLE
Roadmap 1.0

### DISCUSSION THREAD

**Discussion by tailhook** [geldata/gel#1904] [General]
## Goal 1: Stable Language

Parts of language that are stabilized:

1. Query language and data manipulation language (DML)
2. Schema Definition Language (SDL)
3. Data Definition Language (DDL)
4. File formats schema and migrations files
5. Exceptions hierarchy

Stabilization means: all new features are either purely additive (enables
syntax that was not allowed before), or opt-in.

> What can be changed in future:
>
> 1. Any parts of database configuration language
> 2. Introspection, including special syntax (`DESCRIBE`) and introspection
>    objects (`cfg::*` and `sys::*`)
> 3. The underlying language and data structures of migrations

To do:
1. [ ] Disable DML in functions (#1661)
2. [ ] Figure out limitations on generics (#1778)
3. [x] Prohibit implicit cross-product to volatile functions (#1452)
4. [ ] Enforce cardinality on INSERTs (#1973) 
5. [ ] Enforce uniqueness on the (computed) multi links and multi properties

## Goal 2: Reliable Migrations

We should have grades of feature support in migrations:
**stable** features include all addition of objects, fields and functions and
**beta** features include renames, property type changes and so on.

1. [ ] Mark beta-status changes by EdgeDB server
2. [ ] Command-line support of the beta features (maybe even "split the migration")
3. [ ] Documentation: Check-list on how to deal with beta features.

## Goal 3: Production Scenarios Support

1. [x] Postgres connection pooling inside EdgeDB (part of the ["ports-RFC"](#1460))
2. [ ] Postgres replicas support
   * [ ] EdgeDB connection pool changes (#1894)
   * [ ] Protocol update (#1859 + #1896)
   * [ ] Python and TypeScript support for the new protocol issues
3. [ ] End-to-end failover support
   * [ ] Ensure that all language bindings support retry of queries (#1859)
   * [ ] Ensure EdgeDB handles network error and postgres address change gracefully (currently blocked on #1460)
   * [ ] Do end-to-end testing of this feature

## Goal 4: Documentation

1. [ ] More tutorials: how to use EdgeDB with nextjs/nodejs/express etc. How to use EdgeDB with flask/sanic/etc.
2. [ ] Update command-line tool documentation (at least `edgedb server`)
3. [ ] Finalize and host Easy EdgeDB on our website.
4. [ ] Document current error classes

## Goal 5: Command-line tools

1. [ ] Figure out whether we want to restrusture our commands (edgedb/rfcs#14)

**Updates:**
1. Added CLI RFC (edgedb/rfcs#14)
2. Added generics issue (#1778)
3. Added error hierarchy doc and stabilization
4. Remove "Require a FILTER on DELETE and UPDATE  (#1742)" this is going to be a backwards-compatible warning 
5. Add an distinctness and cardinality issue to the language section
6. Add "uniqueness on the (computed) multi links and multi properties" as a separate item

**Comment by 1st1** [comment]
Thanks, Paul. I'll edit your post to add some missing bits.

> Multi-version documentation

This is a non-goal for 1.0 as we'll only have one version. That said, @jaclarke is already made some progress to make this easy to implement.

**Reply by tailhook** [reply]
> Thanks, Paul. I'll edit your post to add some missing bits.

Sure. If there's something significant we may probably want some `Updated:` section below or in a comment. Because it's hard to keep track as we will update it many times probably.

> This is a non-goal for 1.0 as we'll only have one version. That said, @jaclarke is already made some progress to make this easy to implement.

My thought was that we need `1.0` and `master`. But yes, I'm probably overthinking it.

**Reply by 1st1** [reply]
> My thought was that we need 1.0 and master. But yes, I'm probably overthinking it.

I see. I like having "dev" version of the docs (similar to docs.python.org) but yeah, it's not a 1.0 priority item.

**Reply by 1st1** [reply]
> If there's something significant we may probably want some Updated: section below or in a comment. Because it's hard to keep track as we will update it many times probably.

There's edits history for that. But let's see how often we end up updating it.

**Reply by tailhook** [reply]
> There's edits history for that. But let's see how often we end up updating it.

Edit history is useful when we have to investigate some weird edit. It's quite hard to figure out if anything changed by skimming. And reading history entry by entry to find out what have changed (there could be some typo fixes, and so on). But yes let's see.
